File: services/orchestrator/internal/lifecycle/extend.lua
-- extend.lua — đẩy idle-deadline của một session đang chạy về phía trước (B5).
--
-- ⛔ VÌ SAO PHẢI LÀ MỘT SCRIPT, KHÔNG PHẢI ĐỌC-RỒI-GHI BẰNG HAI LỆNH GO:
-- `revision` tồn tại để chặn ĐÚNG cái race này — gateway gia hạn TTL trong khi
-- reaper của orchestrator có thể đang reap cùng session. Đọc revision bằng một
-- lệnh rồi ghi bằng lệnh thứ hai là tự dựng lại khoảng hở giữa hai lệnh, tức là
-- kiểm một giá trị đã cũ. Redis chạy script đơn luồng nên kiểm-và-ghi ở đây
-- không có ai chen vào giữa.
--
-- KEYS[1] = session:{id}        (HASH — SSOT)
-- KEYS[2] = session:{id}:pod    (STRING — con trỏ sống lâu hơn hash, cho reaper)
--
-- ARGV[1] = userId              (chủ sở hữu; server điền từ token đã verify)
-- ARGV[2] = expectedRevision    (0 = bỏ qua kiểm)
-- ARGV[3] = nowUnix
-- ARGV[4] = extendSeconds
-- ARGV[5] = hardCapSeconds      (tính TỪ createdAt, không gia hạn được)
-- ARGV[6] = podPointerGraceSeconds
--
-- Trả: { expiresAt, revision, hardCapReached(0|1), status }
-- error_reply cho mọi nhánh từ chối — Go phân loại bằng tiền tố.

local sess = redis.call('HMGET', KEYS[1],
  'userId', 'status', 'createdAt', 'expiresAt', 'revision')

local userId    = sess[1]
local status    = sess[2]
local createdAt = tonumber(sess[3])
local revision  = tonumber(sess[5]) or 0

-- HMGET trên key không tồn tại trả mảng toàn false — KHÔNG phải lỗi, KHÔNG phải
-- nil reply. Không kiểm ở đây thì "session đã hết hạn" đi tiếp dưới dạng một
-- session rỗng và mọi so sánh authz bên dưới là so với false.
if not userId then
  return redis.error_reply('extend: notfound: session khong ton tai')
end

-- ⛔ CHỦ SỞ HỮU LỆCH TRẢ ĐÚNG MÃ VỚI "KHÔNG TỒN TẠI" (notfound), KHÔNG PHẢI
-- MỘT MÃ RIÊNG. Một mã riêng XÁC NHẬN session đó có thật, biến RPC này thành
-- oracle dò id — cùng lý do GetSession trả NotFound thay vì PermissionDenied.
if userId ~= ARGV[1] then
  return redis.error_reply('extend: notfound: session khong ton tai')
end

-- Trạng thái cuối đời không gia hạn được. Đây là vế thứ hai của thứ mà revision
-- bảo vệ: reaper vừa chuyển session sang EXPIRED/REAPED thì gateway KHÔNG được
-- hồi sinh nó, kể cả khi nó cầm đúng revision cũ.
if status ~= 'CLAIMED' and status ~= 'RUNNING' then
  return redis.error_reply('extend: state: trang thai ' .. tostring(status) .. ' khong cho gia han')
end

local expected = tonumber(ARGV[2]) or 0
if expected ~= 0 and expected ~= revision then
  return redis.error_reply('extend: revision: client thay ' .. expected .. ', hien tai ' .. revision)
end

local now      = tonumber(ARGV[3])
local extend   = tonumber(ARGV[4])
local hardCap  = tonumber(ARGV[5])
local podGrace = tonumber(ARGV[6])

-- HAI ĐỒNG HỒ: idle-deadline gia hạn được, hard cap tính từ createdAt thì không.
-- min() ở đây là thứ khiến RPC này KHÔNG BAO GIỜ là đường giữ pod sống vĩnh
-- viễn: một client bị chiếm quyền spam heartbeat cũng chỉ tới được trần cứng.
local wanted   = now + extend
local capAt    = createdAt + hardCap
local newExpiresAt = wanted
local hardCapReached = 0
if wanted >= capAt then
  newExpiresAt = capAt
  hardCapReached = 1
end

-- Trần cứng đã qua ⇒ không còn gì để gia hạn. Trả về trạng thái hiện tại kèm
-- cờ, để FE hiện "phiên đã tới hạn tối đa" thay vì thấy terminal chết đột ngột.
if newExpiresAt <= now then
  return redis.error_reply('extend: hardcap: da qua tran cung, khong the gia han')
end

revision = revision + 1

redis.call('HSET', KEYS[1],
  'expiresAt', newExpiresAt,
  'lastActiveAt', now,
  'revision', revision)

-- TTL của key phải đi theo expiresAt, nếu không hash biến mất trước hạn (hoặc
-- sống quá hạn) và reaper tầng 1 nghe nhầm thời điểm.
local ttl = newExpiresAt - now
redis.call('EXPIRE', KEYS[1], ttl)
-- Con trỏ pod sống LÂU HƠN hash có chủ ý: lúc `session:{id}` hết hạn thì hash đã
-- biến mất và không còn chỗ nào đọc được podName để xoá pod. Đặt bằng nhau là
-- reaper mù. (Cùng lý do với claim.lua.)
redis.call('EXPIRE', KEYS[2], ttl + podGrace)

return { newExpiresAt, revision, hardCapReached, status }

File: services/orchestrator/internal/lifecycle/extend.go
package lifecycle

import (
	"context"
	_ "embed"
	"errors"
	"fmt"
	"strings"
	"time"

	"github.com/redis/go-redis/v9"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"

	orchestratorv1 "github.com/Nghaiz/DevOps_Learning_Platform/proto/gen/go/orchestrator/v1"
	"github.com/Nghaiz/DevOps_Learning_Platform/services/orchestrator/internal/pool"
	"github.com/Nghaiz/DevOps_Learning_Platform/services/shared/rediskeys"
)

//go:embed extend.lua
var extendLua string

var extendScript = redis.NewScript(extendLua)

// Tiền tố phân loại trong error_reply của extend.lua. Redis không cho script trả
// mã lỗi có cấu trúc, nên đây là kênh duy nhất — giữ khớp KHÍT với file .lua.
const (
	extendErrNotFound = "extend: notfound:"
	extendErrState    = "extend: state:"
	extendErrRevision = "extend: revision:"
	extendErrHardCap  = "extend: hardcap:"
)

// Extend đẩy idle-deadline của một session đang chạy về phía trước.
//
// ⛔ CÔNG THỨC CÓ THỂ KÉO LÙI HẠN, VÀ ĐÓ LÀ ĐIỀU PLAN PIN NGUYÊN VĂN:
// `expires_at = min(now + extend_seconds, created_at + HARD_CAP)` (B5). Hệ quả
// cụ thể: session tạo với SESSION_TTL=1h, tới heartbeat ĐẦU TIÊN với
// EXTEND_DEFAULT=300s thì hạn tụt từ t0+1h xuống t0+~6ph. Nói cách khác,
// SESSION_TTL chỉ là hạn cho tới lần gia hạn đầu; sau đó expires_at luôn bám
// EXTEND_DEFAULT. Điều này có thể ĐÚNG với ý định (đây là idle-deadline, không
// phải tổng thời lượng), nhưng nó va với AC "đóng WS → nối lại cùng {id} trong
// TTL vào đúng pod cũ" — cửa sổ nối lại khi đó chỉ còn EXTEND_DEFAULT chứ không
// phải SESSION_TTL. Hiện thực theo đúng công thức đã pin và ghi lại ở đây;
// muốn hạn chỉ tiến không lùi thì đổi thành max(current, min(...)) — một dòng.
func (s *Service) Extend(
	ctx context.Context, req *orchestratorv1.ExtendSessionRequest,
) (*orchestratorv1.Session, bool, error) {
	sessionID := req.GetSessionId()
	userID := req.GetUserId()

	if userID == "" {
		return nil, false, status.Error(codes.InvalidArgument,
			"user_id bắt buộc — server không suy ra chủ sở hữu")
	}
	if req.GetExtendSeconds() < 0 {
		return nil, false, status.Error(codes.InvalidArgument, "extend_seconds âm")
	}

	sessionKey, err := rediskeys.Session(sessionID)
	if err != nil {
		// Không tách "id sai định dạng" khỏi "không tìm thấy": id tới thẳng từ
		// client, và hai thông báo khác nhau là một oracle bé xíu để dò id.
		return nil, false, status.Error(codes.NotFound, "session không tồn tại")
	}
	sessionPodKey, err := rediskeys.SessionPod(sessionID)
	if err != nil {
		return nil, false, status.Error(codes.NotFound, "session không tồn tại")
	}

	extend := s.cfg.ExtendDefault
	if req.GetExtendSeconds() > 0 {
		extend = time.Duration(req.GetExtendSeconds()) * time.Second
	}

	now := s.now()
	res, err := extendScript.Run(ctx, s.rdb,
		[]string{sessionKey, sessionPodKey},
		userID,
		req.GetExpectedRevision(),
		now.Unix(),
		int64(extend/time.Second),
		int64(s.cfg.HardCap/time.Second),
		// Dùng LẠI hằng của pool, không khai lại: claim.lua và extend.lua đều
		// đặt TTL cho `session:{id}:pod` dài hơn hash đúng khoảng này, và hai
		// giá trị lệch nhau nghĩa là reaper tầng 1 mù ở một trong hai đường.
		int64(pool.PodPointerGrace/time.Second),
	).Slice()
	if err != nil {
		return nil, false, s.mapExtendError(err)
	}

	hardCapReached, err := applyExtendResult(res)
	if err != nil {
		return nil, false, status.Errorf(codes.Internal, "%v", err)
	}

	// Đọc lại để trả về Session ĐẦY ĐỦ. Một lượt Redis nữa trên đường gia hạn
	// (không phải đường nóng của stdin/stdout — G7 gọi cái này theo nhịp traffic
	// chứ không theo từng byte), đổi lấy việc không phải nhân bản logic dựng
	// Session từ giá trị script trả về.
	sess, err := Load(ctx, s.rdb, sessionID)
	if errors.Is(err, ErrSessionNotFound) {
		// Hết hạn đúng giữa script và lượt đọc lại. Hiếm, nhưng có thật.
		return nil, false, status.Error(codes.NotFound, "session không tồn tại")
	}
	if err != nil {
		return nil, false, status.Errorf(codes.Unavailable, "đọc lại session vừa gia hạn: %v", err)
	}

	s.met.ExtendTotal.WithLabelValues("ok").Inc()

	// ⛔ THROTTLE AUDIT: chỉ ghi lần gia hạn CHẠM TRẦN CỨNG.
	//
	// Gia hạn là heartbeat — G7 gọi nó theo nhịp traffic, nên ghi mọi lần sẽ đổ
	// hàng nghìn dòng mỗi phiên vào một bảng nhật ký vòng đời và chôn vùi năm
	// sự kiện thật sự đáng đọc. Câu hỏi "phiên còn sống không" đã có
	// `lastActiveAt` trong Redis trả lời. Lần chạm trần thì khác: nó HIẾM, nó
	// giải thích vì sao phiên sắp chết, và nó là thứ người ta đi tìm khi sinh
	// viên khiếu nại "đang làm thì mất bài".
	if hardCapReached {
		s.audit(ctx, auditEvent{
			SessionID: sess.ID,
			UserID:    sess.UserID,
			Event:     auditEventExtended,
			Tier:      sess.Tier,
			PodName:   sess.PodName,
			Namespace: sess.Namespace,
			Detail:    "chạm trần cứng, không gia hạn thêm được",
		})
	}

	return sess.ToProto(), hardCapReached, nil
}

// applyExtendResult đọc mảng script trả về. Chỉ lấy cờ hard-cap; các giá trị
// khác được đọc lại từ hash để tránh hai nguồn cùng mô tả một session.
func applyExtendResult(res []interface{}) (hardCapReached bool, err error) {
	if len(res) < 3 {
		return false, fmt.Errorf("extend.lua trả %d phần tử, cần >= 3", len(res))
	}
	flag, ok := res[2].(int64)
	if !ok {
		return false, fmt.Errorf("extend.lua trả cờ hard-cap kiểu %T, cần int64", res[2])
	}
	return flag == 1, nil
}

// mapExtendError dịch error_reply của Lua sang mã gRPC.
//
// Mỗi nhánh là một QUYẾT ĐỊNH riêng, không phải bốn cách nói cùng một chuyện:
// NotFound đóng oracle dò id; FailedPrecondition bảo client đọc lại rồi thử lại;
// Unavailable bảo nó thử lại nguyên trạng.
func (s *Service) mapExtendError(err error) error {
	msg := err.Error()
	switch {
	case strings.Contains(msg, extendErrNotFound):
		s.met.ExtendTotal.WithLabelValues("not_found").Inc()
		return status.Error(codes.NotFound, "session không tồn tại")

	case strings.Contains(msg, extendErrRevision):
		// Client cầm revision cũ ⇒ có tiến trình khác vừa ghi. Nó phải đọc lại,
		// XÁC MINH còn đúng chủ + còn sống, rồi mới thử lại — đó là hợp đồng mà
		// G7 hiện thực (thử đúng MỘT lần, vẫn lệch thì đóng 4404).
		s.met.ExtendTotal.WithLabelValues("revision_mismatch").Inc()
		return status.Errorf(codes.FailedPrecondition, "%s", trimLuaPrefix(msg, extendErrRevision))

	case strings.Contains(msg, extendErrState):
		s.met.ExtendTotal.WithLabelValues("bad_state").Inc()
		return status.Errorf(codes.FailedPrecondition, "%s", trimLuaPrefix(msg, extendErrState))

	case strings.Contains(msg, extendErrHardCap):
		s.met.ExtendTotal.WithLabelValues("hard_cap").Inc()
		return status.Error(codes.FailedPrecondition,
			"đã qua trần cứng của phiên, không gia hạn thêm được")

	default:
		s.met.ExtendTotal.WithLabelValues("error").Inc()
		return status.Errorf(codes.Unavailable, "gia hạn session: %v", err)
	}
}

// trimLuaPrefix cắt tiền tố phân loại để thông báo ra ngoài không mang chi tiết
// nội bộ của script.
func trimLuaPrefix(msg, prefix string) string {
	if i := strings.Index(msg, prefix); i >= 0 {
		return strings.TrimSpace(msg[i+len(prefix):])
	}
	return msg
}
